core/store/src/trie/mod.rs

```diff
@@ -63,6 +62,15 @@ enum NodeHandle {
     Hash(CryptoHash),
 }
 
+impl NodeHandle {
+    fn unwrap_hash(&self) -> &CryptoHash {
+        match self {
+            Self::Hash(hash) => hash,
+            Self::InMemory(_) => unreachable!(),
+        }
+    }
```

Comment:
tiny nit, but a more idiomatic version would probably be 

```rust
impl NodeHandle {
    fn as_hash(&self) -> Option<&CryptoHash> {
        match self {
            Self::Hash(hash) => Some(hash),
            Self::InMemory(_) => None,
        }
    }
}
```

that way, we get to re-use Option's `unwrap`.

Reply:
@matklad I guess this way we're losing important knowledge about `unreachable!()` situation.

Reply:
There'll be `.unwrap()` on the call-site, so this just moves the impossibility assertion from the definition site to the call-site. 

There's no much semantic difference between `unreachable!()` and `Option::unwrap` -- both mean that the programmer asserts something at runtime, which can't be explained using rust's type system. 

But, again, this is at most a minor code-style point, in the grand scheme of things, both options are totally fine/

Reply:
@matklad sure. My point in using `unreachable!()` shows that something is impossible by design, but each `unwrap` must be explained why it's safe. It's not about rust, it's about being clear to devs.

(My understanding might be outdated. If it's possible to point that something is not reachable by design another better way (with no using `unreachable!()`) please let me know.)

Reply:
I understand `unreachable!` the same way. But, in my mind, `.unwrap()` also signals impossibility by design? That is, if we look at the call-site,

```rust
hash = *child.unwrap_hash();

/* versus */

hash = *child.as_hash().unwrap();
```

it seems me that both versions say that, by design, `child` is guaranteed to be `Hash` in this particular case. 

The difference is that, when I see `foo.unwrap()`, I immediately know that is signals an impossibility of something. When I see `foo.unwrap_hash()`, I need to check the definition of `unwrap_hash` to realize that it has an unreachable inside.

Reply:
@matklad not sure is it the standard way of understanding in the team. For example, in my latest work, there are explanations of `unwrap` used (https://github.com/near/nearcore/compare/master...gc_headers_2?expand=1#diff-c26b0a777803f79360d3f469862cfcd95deb3c390809095670f6db88d00332f4R236-R239) and I found them useful.

Basically, I work with concepts written by others as black boxes. That's why, in my opinion, `unwrap`s must be explained, otherwise we may end up with lack of understanding what's possible and why. Seeing `unreachable!()` helps me to dive into details much better than guessing can I reuse the same pattern in another place.

One more example - we use tons of `unwrap`s in testing because we don't want to process errors in any other way. It doesn't mean the methods used in tests should never return errors. :)

Reply:
I agree that unwraps should be explained (or, better still, refactored out of existence). But I would also think that, for the same reason, a call to `.unwrap_hash()` should be explained. 

Ie, if I see an uncommitted call to `.unwrap_hash()`, I understand that the author believes, for some good reason, that `unreachable` case won't happen here. But that's exactly the same info I get from seeing `.unwrap()`. That is, `.unwrap()` and `.unwrap_hash()` signal the same level of intentionallyty to me.